nil `*http.Request` panic in `Send*` serializers when the response write fails

**Describe the bug**

We hit a nil-pointer panic in production that traces back to fuego's response
serializers. When a response *write* fails (the client disconnected, or
`WriteTimeout` fired while the handler was still running), fuego falls into its
error-serialization path, and that path ends up calling `SendJSON(w, nil, err)`
with a nil request. `SendJSON` then does `slog.ErrorContext(r.Context(), ...)`
without checking `r`, so `r.Context()` panics.

It's a bit sneaky because the value the handler returned serializes perfectly
fine. The error isn't a marshal error, it's the write to the socket failing. So
a totally normal "client went away" turns into a panic. If you have a recover
middleware you just get a nasty stack trace in the logs every time; if you don't,
the goroutine goes down with it.

Here's the trace we saw on v0.19.0:

```
panic: runtime error: invalid memory address or nil pointer dereference
net/http.(*Request).Context(...)
    /usr/local/go/src/net/http/request.go:353
github.com/go-fuego/fuego.init.func2(...)          serialization.go:164   <- SendJSON, r == nil
github.com/go-fuego/fuego.SendJSONError(...)        serialization.go:218   <- SendJSON(w, nil, err)
github.com/go-fuego/fuego.init.func5(...)           serialization.go:190   <- SendError -> json branch
github.com/go-fuego/fuego.netHttpContext[...].SerializeError  ctx.go:498
github.com/go-fuego/fuego.Flow[...](...)            serve.go:174
```

Same shape exists for `SendXML`, `SendYAML`, and `SendHTML`. They all deref
`r.Context()` without a nil check.

**To Reproduce**

The real-world trigger is just a client disconnecting (or a `WriteTimeout`) mid
response. You don't need a running server to reproduce it though. A
`ResponseWriter` whose `Write` always fails plus a nil request is enough:

```go
type failingWriter struct{ h http.Header }

func (f *failingWriter) Header() http.Header {
	if f.h == nil {
		f.h = http.Header{}
	}
	return f.h
}
func (f *failingWriter) Write([]byte) (int, error) { return 0, errors.New("write: broken pipe") }
func (f *failingWriter) WriteHeader(int)           {}

func TestSendJSON_NilRequest_Panics(t *testing.T) {
	// panics on v0.19.0
	_ = fuego.SendJSON(&failingWriter{}, nil, map[string]string{"ok": "yes"})
}
```

The chain end to end: handler returns a fine value, `Flow` calls
`ctx.Serialize(ans)`, encode fails because the write fails, `Flow` calls
`ctx.SerializeError(err)`, then `SendError` -> `SendJSONError(w, nil, err)` ->
`SendJSON(w, nil, err)` -> `r.Context()` on a nil `r` -> boom.

**Expected behavior**

`SendJSON` and friends shouldn't panic when `r` is nil. If there's no request,
just fall back to `context.Background()` for the log line and return the write
error like normal.

**Framework version (please check if it happens with the last
Fuego version before posting):**

Observed on v0.19.0. I checked `main` too (2026-07-14). The `Send*Error`
functions there now thread the real `r` through instead of passing `nil`, which
fixes the JSON/XML/YAML case from *that* caller. But the serializers themselves
still call `r.Context()` with no nil guard, so it's not actually fixed at the
root:

- `SendHTMLError` still calls `SendHTML(w, nil, ...)`.
- `SendJSON` / `SendXML` / `SendYAML` / `SendHTML` are exported, overridable
  `var`s. Anyone calling them directly with a nil request (custom error
  serializer, re-serializing outside a request scope, etc.) still hits the panic.

So I'd argue the guard belongs in the serializers, not just at the call sites.

**Go version (please check if it happens with the last Go version before posting):**

Not Go-version specific. It's just `(*http.Request).Context()` on a nil receiver.

**Additional context**

The fix is tiny. A little helper:

```go
func requestContext(r *http.Request) context.Context {
	if r == nil {
		return context.Background()
	}
	return r.Context()
}
```

and swap `r.Context()` for `requestContext(r)` at the four deref sites (`SendJSON`
~L164, `SendXML` ~L228, `SendYAML` ~L138, `SendHTML` ~L265 on the `CtxRenderer`
path). No signature changes, behavior is identical when `r != nil`, so it's fully
backward compatible. Keeping the `main` change that threads `r` through is still
nice for keeping trace/request IDs in the failure logs; the guard is just belt
and suspenders on top.

For now we're working around it by overriding `fuego.SendJSON` at startup with a
nil-safe copy, but that only covers the JSON path and duplicates your logic, so
I'd rather see it fixed upstream.

--- a/serialization.go
+++ b/serialization.go
@@ -79,6 +79,13 @@ func transformOut[T any](ctx context.Context, ans T) (T, error) {
 	return ans, nil
 }
 
+func requestContext(r *http.Request) context.Context {
+	if r == nil {
+		return context.Background()
+	}
+	return r.Context()
+}
+
 type Sender func(http.ResponseWriter, *http.Request, any) error
 
 // Send sends a response.
@@ -135,7 +142,7 @@ var SendYAML = func(w http.ResponseWriter, r *http.Request, ans any) (err error)
 	err = yaml.NewEncoder(w).Encode(ans)
 	if err != nil {
 		w.WriteHeader(http.StatusInternalServerError)
-		slog.ErrorContext(r.Context(), "Cannot serialize returned response to YAML", "error", err)
+		slog.ErrorContext(requestContext(r), "Cannot serialize returned response to YAML", "error", err)
 		_, _ = w.Write([]byte(`{"error":"Cannot serialize returned response to YAML"}`))
 	}
 	return err
@@ -161,7 +168,7 @@ var SendJSON = func(w http.ResponseWriter, r *http.Request, ans any) error {
 	w.Header().Set("Content-Type", "application/json")
 	err := json.NewEncoder(w).Encode(ans)
 	if err != nil {
-		slog.ErrorContext(r.Context(), "Cannot serialize returned response to JSON", "error", err, "errtype", fmt.Sprintf("%T", err))
+		slog.ErrorContext(requestContext(r), "Cannot serialize returned response to JSON", "error", err, "errtype", fmt.Sprintf("%T", err))
 		var unsupportedType *json.UnsupportedTypeError
 		if errors.As(err, &unsupportedType) {
 			return NotAcceptableError{
@@ -225,7 +232,7 @@ var SendXML = func(w http.ResponseWriter, r *http.Request, ans any) error {
 	w.Header().Set("Content-Type", "application/xml")
 	err := xml.NewEncoder(w).Encode(ans)
 	if err != nil {
-		slog.ErrorContext(r.Context(), "Cannot serialize returned response to XML", "error", err, "errtype", fmt.Sprintf("%T", err))
+		slog.ErrorContext(requestContext(r), "Cannot serialize returned response to XML", "error", err, "errtype", fmt.Sprintf("%T", err))
 		var unsupportedType *xml.UnsupportedTypeError
 		if errors.As(err, &unsupportedType) {
 			return NotAcceptableError{
@@ -250,7 +257,7 @@ func SendXMLError(w http.ResponseWriter, r *http.Request, err error) {
 	w.WriteHeader(status)
 	err = SendXML(w, r, err)
 	if err != nil {
-		slog.ErrorContext(r.Context(), "Cannot serialize returned response to XML", "error", err)
+		slog.ErrorContext(requestContext(r), "Cannot serialize returned response to XML", "error", err)
 		_, _ = w.Write([]byte(`{"error":"Cannot serialize returned response to XML"}`))
 	}
 }
@@ -262,7 +269,7 @@ var SendHTML = func(w http.ResponseWriter, r *http.Request, ans any) error {
 
 	ctxRenderer, ok := ans.(CtxRenderer)
 	if ok {
-		return ctxRenderer.Render(r.Context(), w)
+		return ctxRenderer.Render(requestContext(r), w)
 	}
 
 	renderer, ok := ans.(Renderer)

--- a/serialization_test.go
+++ b/serialization_test.go
@@ -667,3 +667,25 @@ func TestSendError(t *testing.T) {
 		})
 	}
 }
+
+func TestSend_NilRequest_WriteError_NoPanic(t *testing.T) {
+	cases := map[string]func(http.ResponseWriter, *http.Request, any) error{
+		"json": SendJSON,
+		"xml":  SendXML,
+		"yaml": SendYAML,
+	}
+	for name, send := range cases {
+		t.Run(name, func(t *testing.T) {
+			require.NotPanics(t, func() {
+				err := send(&errorWriter{}, nil, response{Message: "Hello World", Code: http.StatusOK})
+				require.Error(t, err, "expected the underlying write error to be returned")
+			})
+		})
+	}
+
+	t.Run("html CtxRenderer", func(t *testing.T) {
+		require.NotPanics(t, func() {
+			_ = SendHTML(httptest.NewRecorder(), nil, RenderString("Hello World"))
+		})
+	})
+}

